server/manager/src/main/java/org/apache/accumulo/master/Master.java

```diff
@@ -852,34 +866,57 @@ private void checkForHeldServer(SortedMap<TServerInstance,TabletServerStatus> ts
     }
 
     private long balanceTablets() {
-      List<TabletMigration> migrationsOut = new ArrayList<>();
-      long wait = tabletBalancer.balance(Collections.unmodifiableSortedMap(tserverStatus),
-          migrationsSnapshot(), migrationsOut);
-
-      for (TabletMigration m : TabletBalancer.checkMigrationSanity(tserverStatus.keySet(),
-          migrationsOut)) {
-        if (migrations.containsKey(m.tablet)) {
+      BalanceParamsImpl params = BalanceParamsImpl.fromThrift(tserverStatusForBalancer,
+          tserverStatus, migrationsSnapshot());
+      long wait = tabletBalancer.balance(params);
+
+      for (TabletMigration m : checkMigrationSanity(tserverStatusForBalancer.keySet(),
+          params.migrationsOut())) {
+        KeyExtent ke = TabletIdImpl.toKeyExtent(m.getTablet());
+        if (migrations.containsKey(ke)) {
           log.warn("balancer requested migration more than once, skipping {}", m);
           continue;
         }
-        migrations.put(m.tablet, m.newServer);
+        TServerInstance tserverInstance = TabletServerIdImpl.toThrift(m.getNewTabletServer());
+        migrations.put(ke, tserverInstance);
         log.debug("migration {}", m);
       }
-      if (migrationsOut.isEmpty()) {
+      if (params.migrationsOut().isEmpty()) {
         synchronized (balancedNotifier) {
           balancedNotifier.notifyAll();
         }
       } else {
-        nextEvent.event("Migrating %d more tablets, %d total", migrationsOut.size(),
+        nextEvent.event("Migrating %d more tablets, %d total", params.migrationsOut().size(),
             migrations.size());
       }
       return wait;
     }
 
+    private List<TabletMigration> checkMigrationSanity(Set<TabletServerId> current,
+        List<TabletMigration> migrations) {
+      return migrations.stream().filter(m -> {
+        boolean includeMigration = false;
+        if (m.getTablet() == null) {
+          log.warn("Balancer gave back a null tablet {}", m);
+        } else if (m.getNewTabletServer() == null) {
+          log.warn("Balancer did not set the destination {}", m);
+        } else if (m.getOldTabletServer() == null) {
+          log.warn("Balancer did not set the source {}", m);
+        } else if (!current.contains(m.getOldTabletServer())) {
+          log.warn("Balancer wants to move a tablet from a server that is not current: {}", m);
+        } else if (!current.contains(m.getNewTabletServer())) {
+          log.warn("Balancer wants to move a tablet to a server that is not current: {}", m);
```

Comment:
Most of these seem like they should be errors instead of warnings, as they are incorrectly behaving balancers. I think the "wants to move... not current" are probably okay as warnings, though, because they seem like they could happen as the result of things changing on the cluster since the balancer made a choice.

Reply:
This is just a copy of the checkMigrationSanity method from the now deprecated [TabletBalancer class](https://github.com/apache/accumulo/blob/main/server/base/src/main/java/org/apache/accumulo/server/master/balancer/TabletBalancer.java#L252). I added it here since this was the only place it was used.

I think it makes sense to change the level to error, though. I can certainly do that.